docs/DEVELOPING.md

```diff
@@ -124,6 +127,227 @@ Take a look at the [MongoDB connector](../connectors/sources/mongo.py) for more
+
+##### When should failure responses be retried?
+
+As discussed above, many APIs utilize rate limits and will issue 429 responses for requests that should be retried later.
+However, sometimes there are other cases that warrant retries. 
+
+Some services experience frequent-but-transient timeout issues.
+Some services experiencing a large volume of write operations may sometimes respond with 409 "conflict" codes.
+Sometimes downloading large files encounter packet loss, and while the response is a 200 OK, the checksum of the downloaded file does not match.
+
+All of these are operations that may make sense to retry a time or two.
+
+A note on retries - do not retry infinitely, and use backoff.
+A service may block your connector or specific features if you retry too aggressively.
+
+##### What is the permission model used by the 3rd-party?
+
+Whether or not you plan to support Document Level Security with your connector, it is important to understand how the 3rd-party system protects data and restricts access.
+This allows you to ensure that synced documents contain the relevant metadata to indicate ownership/permissions so that any consumer of your connector can choose to implement their own DLS if they so desire.
+
+For Sharepoint Online, for example, this involves syncing the groups and users that own a document or to whom a document has been shared to.
+Sometimes these ownership models can be complex.
+For Sharepoint Online, you must consider:
+- the individual creator of the document
+- the "owners" of any site where the document lives
+- the "members" of any site where the document lives
+- the "visitors" of any site where the document lives
+- Any Office365 groups that this document has been shared with
+- Any Sharepoint groups that this document has been shared with
+- Any Azure AD groups that this document has been shared with
+- Any individuals (email addresses) that this document has been shared with
+
+How this information is serialized onto a document is less important than ensuring that it _is_ included.
+One common pattern is to put all permissions information in an `_allow_permissions` field with an array of text values.
+But it would be equally valid to split each of the above bullets into their own fields.
+Or to have two fields, one for "individuals" and one for "groups".
+
+#### What is the expected performance for the connector?
+
+There are no hard-and-fast rules here, unless you're building a connector with a specific customer in mind, who has specific requirements.
+However, it is good to ensure that any connector you build has comparable performance to that of other connectors.
+To measure this, you can simply utilize the `make ftest` (functional tests) and look at the generated performance dashboards to compare apples-to-apples.
+Pay particular attention to memory, CPU, and file handle metrics.
+
+See [Integration tests](#integration-tests) for more details.
+
+##### How much memory should it require?
```

Comment:
Might be a good idea to add a note about how the connectors keep documents' IDs in memory and how this can impact memory consumption. 

Like `if your index has 100k documents and one ID is 10kb then you need 100k*10=1m bytes~1mb extra.` This might be useful for connectors that store long IDs.

Reply:
Good point! I'll capture that.